src/daemon/priv-linux.c

```diff
@@ -54,28 +55,88 @@ priv_open(char *file)
 }
 
 static int
-asroot_ethtool_real(const char *ifname, struct ethtool_cmd *ethc) {
+asroot_ethtool_real(const char *ifname, struct ethtool_link_usettings *uset) {
 	int rc, sock = -1;
-	struct ifreq ifr = {
-		.ifr_data = (caddr_t)ethc
-	};
+	struct ifreq ifr = {};
+	if ((rc = sock = socket(AF_INET, SOCK_DGRAM, 0)) == -1) {
+		return rc;
+	}
 	strlcpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
 
-	memset(ethc, 0, sizeof(struct ethtool_cmd));
-	ethc->cmd = ETHTOOL_GSET;
-	if (((rc = sock = socket(AF_INET, SOCK_DGRAM, 0)) == -1) ||
-	    (rc = ioctl(sock, SIOCETHTOOL, &ifr)) != 0) {
-		if (sock != -1) close(sock);
-		return rc;
+	/* Try with ETHTOOL_GLINKSETTINGS first */
+	struct {
+		struct ethtool_link_settings req;
+		uint32_t link_mode_data[3 * ETHTOOL_LINK_MODE_MASK_MAX_KERNEL_NU32];
+	} ecmd;
+	static int8_t nwords = 0;
+
+	if (nwords == 0) {
+		/* Do a handshake first. We assume that this is device-independant. */
+		memset(&ecmd, 0, sizeof(ecmd));
+		ecmd.req.cmd = ETHTOOL_GLINKSETTINGS;
+		ifr.ifr_data = (caddr_t)&ecmd;
+		rc = ioctl(sock, SIOCETHTOOL, ifr);
+		if (rc == 0) {
+			nwords = -ecmd.req.link_mode_masks_nwords;
+			log_debug("privsep", "glinksettings nwords is %" PRId8, nwords);
+		}
 	}
+	if (nwords > 0) {
+		memset(&ecmd, 0, sizeof(ecmd));
+		ecmd.req.cmd = ETHTOOL_GLINKSETTINGS;
```

Comment:
I don't have a network card using `ETHTOOL_GLINKSETTINGS`, so this path is untested. I see the ability of using `ETHTOOL_GLINKSETTINGS` as non-root has been added in 4.9-rc7. I have asked if it would be possible to also push the patch to stable branches. In the meantime, I run this as root now. Could you test?